collections/src/main/scala/strawman/collection/immutable/ChampHashSet.scala

```diff
@@ -0,0 +1,607 @@
+package strawman
+package collection
+package immutable
+
+import mutable.{Builder, ImmutableBuilder}
+import Hashing.computeHash
+
+import scala.{Any, AnyRef, Array, Boolean, Double, IndexOutOfBoundsException, Int, NoSuchElementException, SerialVersionUID, Serializable, Unit, `inline`, sys, Some, None}
+import scala.Predef.{assert, require, ???}
+
+import java.lang.Integer.bitCount
+import java.lang.Math.ceil
+import java.lang.System.arraycopy
+
+/** This class implements immutable sets using a Compressed Hash-Array Mapped Prefix-tree.
+  * See paper https://michael.steindorfer.name/publications/oopsla15.pdf for more details.
+  *
+  *  @tparam A      the type of the elements contained in this hash set.
+  *
+  *  @author  Michael J. Steindorfer
+  *  @version 2.13
+  *  @since   2.13
+  *  @define Coll `immutable.ChampHashSet`
+  *  @define coll immutable champ hash set
+  */
+@SerialVersionUID(2L)
+private[immutable] sealed case class ChampHashSet[A](val rootNode: SetNode[A], val cachedJavaHashCode: Int, val cachedSize: Int)
```

Comment:
Here too, you don’t want to use a “case” class, just a plain class. It doesn’t have to be sealed but you can make it `final` if you want. And the type should not be private, although you can make the primary constructor private:

~~~ scala
final class ChampHashSet[A] private[immutable] (val rootNode: SetNode[A], val cachedJavaHashCode: Int, val cachedSize: Int)
~~~

Reply:
Done.